spec.html

```diff
@@ -25410,24 +25483,35 @@ <h1>Evaluate ( )</h1>
             <dd>a Cyclic Module Record _module_</dd>
 
             <dt>description</dt>
-            <dd>Evaluate transitions this module's [[Status]] from ~linked~ to ~evaluated~. If execution results in an exception, that exception is recorded in the [[EvaluationError]] field and rethrown by future invocations of Evaluate. (Most of the work is done by the auxiliary function InnerModuleEvaluation.)</dd>
+            <dd>Evaluate transitions this module's [[Status]] from ~linked~ to either ~evaluating-async~ or ~evaluated~. The Promise returned by Evaluate is allocated by the first invocation of Evaluate, and its capability is stored in the [[TopLevelCapability]] field. If execution results in an exception, the Promise is rejected. Future invocations of Evaluate return the same Promise. (Most of the work is done by the auxiliary function InnerModuleEvaluation.)</dd>
           </dl>
 
           <emu-alg>
             1. Assert: This call to Evaluate is not happening at the same time as another call to Evaluate within the surrounding agent.
-            1. Assert: _module_.[[Status]] is ~linked~ or ~evaluated~.
+            1. Assert: _module_.[[Status]] is ~linked~, ~evaluating-async~, or ~evaluated~.
+            1. If _module_.[[Status]] is ~evaluating-async~ or ~evaluated~, set _module_ to _module_.[[CycleRoot]].
```

Comment:
I *think* @bakkot is right that with his suggestion the subsequent step becomes unreachable. Namely, the only modules that have [[TopLevelCapability]] set are the cycle roots, which are also the modules that had `Evaluate()` invoked. Since Evaluate() asserts that it's not reentrant, so if `_module_.[[Status]] is ~evaluating-async~ or ~evaluated~`, then [[CycleRoot]] should be set, and the cycle root module must have a [[TopLevelCapability]]. Finally, the cycle root module has [[CycleRoot]] pointing to itself.

Reply:
`[[TopLevelCapability]]` is not specifically about cycles but more about a top-level `Evaluate()` call being performed which represents a whole static graph execution that is then stored as this promise. It is set on all modules for which `Evaluate` has been called. And it is only set on one module for each `Evaluate()` call. Many cycle roots will not have `[[TopLevelCapability]]` set.

> (If not, I am very confused.)

Happy to clarify :)

Reply:
Wait, I'm confused now. Do you have a concrete example where we call Evaluate() on a module _m_, and _m_.[[CycleRoot]] doesn't have a top-level capability? That would create a new promise right now, which seems weird?

Reply:
For eg A -> B -> C -> B, where you have a top-level import or dynamic import of A, then A.[[TopLevelCapability]] would be defined as the top-level `Evaluate()` operation promise, while B and C would have no top-level capability as they pass through `InnerModuleEvaluation` not the top level evaluation. If you then performed a dynamic import to B or C during this process, only then would the TopLevelCapability be defined as a top-level evaluation promise. And the promise would be based on attaching a completion handler to the cycle root through the completion callback mechanism.

Reply:
This top-level capability is effectively just a way to queue up and share multiple top-level evaluation promises - it could equally well be replaced with a top-level evaluate callback mechanism that similarly handles evaluation completion (and that would be a spec structure that could also in turn permit sync evaluation for sync graphs fwiw).

Reply:
> A -> B -> C -> B, where you have a top-level import or dynamic import of A, then A.[[TopLevelCapability]] would be defined as the top-level Evaluate() operation promise, while B and C would have no top-level capability as they pass through InnerModuleEvaluation not the top level evaluation.

Ah okay. Might be hard to capture in a note... hm, not sure how to best capture this so future readers don't have the same confusion.

Reply:
Yup. I [tweaked](https://github.com/tc39/ecma262/pull/2408/commits/5a0edba0501f00b069659ddeae82d151cf1fef36) the descriptions of [[CycleRoot]] and [[TopLevelCapability]] (and also of [Evaluate](https://github.com/tc39/ecma262/pull/2408/commits/9a4743005f4caeee64d0fd213c1528305e5ac8e1)) in a way which will hopefully make this clearer for future readers.